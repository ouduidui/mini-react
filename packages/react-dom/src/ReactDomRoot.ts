import type { Container } from 'react-reconciler'
import { createContainer, updateContainer } from 'react-reconciler'
import type { ReactElement } from 'shared/index'

export function createRoot(container: Container) {
  const root = createContainer(container)
  return {
    render(element: ReactElement) {
      updateContainer(element, root)
    },
  }
}
