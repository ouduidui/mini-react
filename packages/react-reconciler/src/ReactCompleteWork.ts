import type { FiberNode } from './ReactFiber'
import { NoFlags } from './ReactFiberFlags'
import type { Instance } from './ReactFiberHostConfig'
import { appendInitialChild, createInstance } from './ReactFiberHostConfig'
import { WorkTag } from './ReactWorkTags'

export const completeWork = (workInProgress: FiberNode) => {
  switch (workInProgress.tag) {
    case WorkTag.HostRoot:
      bubbleProperties(workInProgress)
      return null
    case WorkTag.HostComponent:
      workInProgress.stateNode = createInstance(workInProgress.type)
      appendAllChildren(workInProgress.stateNode, workInProgress)

      bubbleProperties(workInProgress)
      return null
    default:
      return null
  }
}

function bubbleProperties(completeWork: FiberNode) {
  let subtreeFlags = NoFlags
  let child = completeWork.child
  while (child !== null) {
    subtreeFlags |= child.subtreeFlags
    subtreeFlags |= child.flags

    child.return = completeWork
    child = child.sibling
  }
  completeWork.subtreeFlags |= subtreeFlags
}

function appendAllChildren(parent: Instance, workInProgress: FiberNode) {
  // 遍历workInProgress所有子孙 DOM元素，依次挂载
  let node = workInProgress.child
  while (node) {
    if (node.tag === WorkTag.HostComponent) {
      appendInitialChild(parent, node.stateNode)
    }
    else if (node.child) {
      node.child.return = node
      node = node.child
      continue
    }

    if (node === workInProgress)
      return

    while (node.sibling) {
      if (!node.return || node.return === workInProgress)
        return

      node = node.return
    }

    node.sibling!.return = node.return
    node = node.sibling
  }
}
