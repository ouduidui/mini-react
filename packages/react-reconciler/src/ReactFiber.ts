
import type { ReactElement } from 'shared/index'
import type { Key, Props } from 'shared/index'
import type { UpdateQueue } from './ReactFiberClassUpdateQueue'
import type { Flags } from './ReactFiberFlags'
import { NoFlags } from './ReactFiberFlags'
import type { Container } from './ReactFiberHostConfig'
import { WorkTag } from './ReactWorkTags'

export class FiberNode {
  tag: WorkTag
  key: Key
  type: any
  flags: Flags = NoFlags
  subtreeFlags: Flags = NoFlags
  pendingProps: Props
  memoizedProps: Props | null = null
  stateNode: any = null

  return: FiberNode | null = null
  sibling: FiberNode | null = null
  child: FiberNode | null = null
  alternate: FiberNode | null = null
  index = 0

  updateQueue: UpdateQueue | null = null
  memoizedState: any

  constructor(tag: WorkTag, pendingProps: Props, key: Key) {
    this.tag = tag
    this.key = key
    this.pendingProps = pendingProps
  }
}

export class FiberRootNode {
  constructor(public container: Container, public current: FiberNode) {
    current.stateNode = this
  }
}

export const createWorkInProgress = (current: FiberNode, pendingProps: Props) => {
  let wip = current.alternate

  if (!wip) {
    // mount
    wip = new FiberNode(current.tag, pendingProps, current.key)
    wip.type = current.type
    wip.stateNode = current.stateNode

    wip.alternate = current
    current.alternate = wip
  }
  else {
    // update
    wip.pendingProps = pendingProps
  }

  wip.updateQueue = current.updateQueue
  wip.flags = current.flags
  wip.child = current.child

  // 数据
  wip.memoizedProps = current.memoizedProps
  wip.memoizedState = current.memoizedState

  return wip
}

export function createFiberFromElement(element: ReactElement) {
  const { type, key, props } = element
  let fiberTag = WorkTag.FunctionComponent

  if (typeof type === 'string')
    fiberTag = WorkTag.HostComponent

  const fiber = new FiberNode(fiberTag, props, key)
  fiber.type = type

  return fiber
}
