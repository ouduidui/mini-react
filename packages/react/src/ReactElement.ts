import hasOwnProperty from 'shared/hasOwnProperty'
import { REACT_ELEMENT_TYPE } from 'shared/ReactSymbol'
import type { Key } from 'shared/ReactTypes'

interface ReactElement {
  /**
   * https://overreacted.io/zh-hans/why-do-react-elements-have-typeof-property/
   */
  $$typeof: Symbol
  type: any
  key: Key | null
  props: any
}

/**
 * 保留属性，以下属性不会加入到props中
 */
const RESERVED_PROPS = {
  key: true,
  ref: true,
  __self: true,
  __source: true,
}

export function createElement(
  type: any,
  config?: Record<string, any>,
  ...children: any[]
): ReactElement {
  const props: Record<string, any> = {}

  let key: Key | null = null

  if (config) {
    for (const propName in config) {
      if (
        hasOwnProperty.call(config, propName)
        && !hasOwnProperty.call(RESERVED_PROPS, propName)
      )
        props[propName] = config[propName]
    }
  }

  if (type?.defaultProps) {
    const defaultProps = type.defaultProps
    for (const propName in defaultProps) {
      if (props[propName] === undefined)
        props[propName] = defaultProps[propName]
    }
  }

  if (config?.key)
    key = `${config?.key}`

  if (children.length === 1)
    props.children = children[0]
  else if (children.length > 1)
    props.children = children

  return {
    $$typeof: REACT_ELEMENT_TYPE,
    type,
    key,
    props,
  }
}
