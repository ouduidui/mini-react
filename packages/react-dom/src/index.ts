import { createRoot } from './ReactDomRoot'

export {
  createRoot,
}
