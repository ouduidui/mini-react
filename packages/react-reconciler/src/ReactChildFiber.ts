import type { ReactElement } from 'shared/index'
import { REACT_ELEMENT_TYPE } from 'shared/index'
import type { FiberNode } from './ReactFiber'
import { createFiberFromElement } from './ReactFiber'
import { Placement } from './ReactFiberFlags'

function createChildReconciler(shouldTrackEffect: boolean) {
  function reconcileSingleElement(
    returnFiber: FiberNode,
    currentFirstChild: FiberNode | null,
    newChild: ReactElement,
  ) {
    const fiber = createFiberFromElement(newChild)
    fiber.return = returnFiber
    return fiber
  }

  function placeSingleChild(fiber: FiberNode) {
    if (shouldTrackEffect)
      fiber.flags |= Placement

    return fiber
  }

  function reconcileChildFibers(
    returnFiber: FiberNode,
    currentFirstChild: FiberNode | null,
    newChild?: ReactElement,
  ) {
    if (typeof newChild === 'object' && newChild) {
      switch (newChild.$$typeof) {
        case REACT_ELEMENT_TYPE:
          return placeSingleChild(
            reconcileSingleElement(returnFiber, currentFirstChild, newChild),
          )
        default:
          return null
      }
    }
    return null
  }

  return reconcileChildFibers
}

export const reconcileChildFibers = createChildReconciler(true)
export const mountChildFibers = createChildReconciler(false)
