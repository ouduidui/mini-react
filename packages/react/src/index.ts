import { createElement } from './ReactElement'

export {
  createElement,
}
