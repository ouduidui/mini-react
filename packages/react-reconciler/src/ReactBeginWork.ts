import type { ReactElement } from 'shared/dist'
import { mountChildFibers, reconcileChildFibers } from './ReactChildFiber'
import type { FiberNode } from './ReactFiber'
import { WorkTag } from './ReactWorkTags'

export const beginWork = (workInProgress: FiberNode) => {
  switch (workInProgress.tag) {
    case WorkTag.HostRoot:
      return updateHostRoot(workInProgress)
    default :
      return null
  }
}

function updateHostRoot(workInProgress: FiberNode) {
  const nextProps = workInProgress.pendingProps
  const nextChildren = nextProps.children
  reconcileChildren(workInProgress, nextChildren)
  return workInProgress.child
}

function reconcileChildren(workInProgress: FiberNode, children?: ReactElement) {
  const current = workInProgress.alternate

  // update
  if (current)
    workInProgress.child = reconcileChildFibers(workInProgress, current.child, children)

  // mount
  else
    workInProgress.child = mountChildFibers(workInProgress, null, children)
}
