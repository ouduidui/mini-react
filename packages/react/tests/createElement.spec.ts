import { describe, expect, it } from 'vitest'
import { createElement } from '../src/ReactElement'

describe('createElement', () => {
  it('should create a ReactElement', () => {
    expect(createElement('div')).toMatchSnapshot()
  })

  it('has props', () => {
    const elm = createElement('div', { className: 'm-10', id: 'divEl', key: 123 })
    expect(elm).toMatchSnapshot()
    expect(elm.key).toBe('123')
  })
})
