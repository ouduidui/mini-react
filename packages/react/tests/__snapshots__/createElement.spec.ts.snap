// Vitest Snapshot v1

exports[`createElement > has props 1`] = `
<div
  className="m-10"
  id="divEl"
/>
`;

exports[`createElement > should create a ReactElement 1`] = `<div />`;
